File: Solutions/Solution380.java
class RandomizedSet {
  Map<Integer, Integer> map;
  List<Integer> existingElementIndices;

  public RandomizedSet() {
    this.map = new HashMap<>();
    this.existingElementIndices = new ArrayList<>();
  }

  public boolean insert(int val) {
    if (map.containsKey(val)) {
      return false;
    } else {
      map.put(val, existingElementIndices.size());
      existingElementIndices.add(val);
      return true;
    }
  }

  public boolean remove(int val) {
    if (!map.containsKey(val)) {
      return false;
    } else {
      int index = map.get(val);
      int lastVal = existingElementIndices.get(existingElementIndices.size() - 1);

      // Update the index of the last element in the map
      map.put(lastVal, index);

      // Swap the values at index and size-1, then remove the last element
      existingElementIndices.set(index, lastVal);
      existingElementIndices.remove(existingElementIndices.size() - 1);

      map.remove(val);
      return true;
    }
  }

  public int getRandom() {
    Random rand = new Random();
    return existingElementIndices.get(rand.nextInt(existingElementIndices.size()));
  }
}

/**
 * Your RandomizedSet object will be instantiated and called as such: RandomizedSet obj = new
 * RandomizedSet(); boolean param_1 = obj.insert(val); boolean param_2 = obj.remove(val); int
 * param_3 = obj.getRandom();
 */

File: Solutions/Solution380.cpp
class RandomizedSet
{
    std::vector<int> valueArray;
    std::map<int, int> indexMap;

public:
    RandomizedSet() {}

    bool insert(int val)
    {
        if (indexMap.contains(val))
        {
            return false;
        }
        else
        {
            indexMap[val] = valueArray.size();
            valueArray.push_back(val);
            return true;
        }
    }

    bool remove(int val)
    {
        if (!indexMap.contains(val))
        {
            return false;
        }
        else
        {
            // Swap to the end
            int index = indexMap[val];

            valueArray[index] = valueArray.back();
            indexMap[valueArray.back()] = index;

            // Remove last element
            valueArray.pop_back();
            indexMap.erase(val);
            return true;
        }
    }

    int getRandom()
    {
        int random = rand() % valueArray.size();
        return valueArray[random];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */